🌱 Use structured rancher proxy types

Danil-Grigorev opened the pull request:
<!--
Label the PR with the kind of change this for:

kind/feature
kind/bug
kind/documentation
kind/regression
kind/*
-->

kind/refactor

**What this PR does / why we need it**:

While using controller runtime a preferred way to handle CRDs in envtest is by supplying a `CRDDirectoryPaths` with CRD definitions needed for registering API type in api server (See: https://book.kubebuilder.io/reference/envtest.html?highlight=CRDDirectoryPaths#writing-tests). We can leverage that and remove the need for unstructured objects for rancher types without needing to import it from rancher/rancher.

This will make rancher proxy types compatible with client.Object interface unlocking a use for
something like [komega](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest/komega) for testing and simplify transition to rancher API types in the future, when it will be ready to do so.

E2E tests: https://github.com/Danil-Grigorev/rancher-turtles/actions/runs/6089148822

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [ ] squashed commits into logical changes
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

alexander-demicev commented on Makefile:
```diff
@@ -140,12 +140,17 @@ help: ## Display this help.
-		rbac:roleName=manager-role
+manifests: vendor controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
+	$(CONTROLLER_GEN) rbac:roleName=manager-role crd paths="./..." output:crd:artifacts:config=hack/crd/bases
+	$(CONTROLLER_GEN) rbac:roleName=manager-role crd paths="./vendor/sigs.k8s.io/cluster-api/..." output:crd:artifacts:config=hack/crd/bases
```
I'm not a fan of using vendor if it's possible to avoid it, generating CRDs in vendor also adds some complexity to development

furkatgofurov7 replied:
+1, vendoring bloats the repo a lot from my experience and technically should not be needed, at least for now IMO

Danil-Grigorev replied:
It is being ignored by the gitignore, so it is not being used in the build process and won’t bloat the repo. Also this is not a task you would manually run for doing any commands, it is only triggered by generate tasks. Only thing it does here, is providing a relative directory for the CRDs to be picked up from. This is workflow wich worked perfectly for me in the past and allows us to avoid manually writing boilerplate code where possible. Additionally, when (if) we will be able to import rancher cluster from the rancher repo, there will be no business logic changes thanks to this refactor and centralized use of client.Object interface compatible structures.

richardcase replied:
As its ignored are we happy to resolve this?

alexander-demicev reviewed (commented):
not sure how others feel about it but IMO it seems to add more complexity than wrappers around rancher objects

furkatgofurov7 commented:
@Danil-Grigorev this needs a rebase, I'll take a look once again after that

Danil-Grigorev commented:
I addressed some feedback on this PR, added removal of vendor directory to emphasize that we don’t use vendor for anything else then manifest generation (and is used only due to being a known relative location). Rebased the PR, updated e2e tests with the use of komega. Example CI runs are here: https://github.com/Danil-Grigorev/rancher-turtles/actions/runs/6089148822

furkatgofurov7 commented on internal/rancher/management/v3/clusterregistrationtoken.go:
```diff
@@ -0,0 +1,33 @@
+package v3
```
is there reason for it to be v3?

alexander-demicev replied:
I guess Rancher objects are v3 so that is done to align with them

furkatgofurov7 commented on Makefile:
```diff
@@ -174,12 +174,23 @@ help: ## Display this help.
-		paths=./internal/controllers \
-		rbac:roleName=manager-role
+manifests: vendor controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
+	$(CONTROLLER_GEN) rbac:roleName=manager-role crd paths="./..." output:crd:artifacts:config=hack/crd/bases
```
not sure why just`crd` is included that way, maybe you meant:

```suggestion
	$(CONTROLLER_GEN) \ 
	        paths="./..." output:crd:artifacts:config=hack/crd/bases \
	        crd:crdVersions=v1 \
	        rbac:roleName=manager-role
```

also, should we have a new Makefile target, something like `make-clean-manifests` to run at the beginning of this to ensure we remove the previously generated manifests?

Danil-Grigorev replied:
It is not only v1 we are using. It is already being overwritten on each `make manifests` makefile run.

furkatgofurov7 replied:
That I know, I was thinking of having a clean repo before running it, however it is personal choice I guess

furkatgofurov7 commented on internal/controllers/import_controller.go:
```diff
@@ -193,32 +188,34 @@ func (r *CAPIImportReconciler) Reconcile(ctx context.Context, req ctrl.Request)
-	rancherClusterName := turtelesnaming.Name(capiCluster.Name).ToRancherName()
+	rancherCluster := &provisioningv1.Cluster{ObjectMeta: metav1.ObjectMeta{
+		Namespace: capiCluster.Namespace,
+		Name:      turtelesnaming.Name(capiCluster.Name).ToRancherName(),
```
this needs to be fixed for the import of package itself as well:

```suggestion
		Name:      turtlesnaming.Name(capiCluster.Name).ToRancherName(),
```

Danil-Grigorev replied:
Looking at this line, I really asking myself why I did it like this 😔

furkatgofurov7 commented on internal/controllers/import_controller.go:
```diff
@@ -193,32 +188,34 @@ func (r *CAPIImportReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 func (r *CAPIImportReconciler) reconcile(ctx context.Context, capiCluster *clusterv1.Cluster) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
 	// fetch the rancher clusters
```
```suggestion
	// fetch the rancher cluster
```

furkatgofurov7 commented on internal/rancher/management/v3/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+limitations under the License.
+*/
+
+// Package v1alpha3 contains API Schema definitions for the clusterctl v1alpha3 API group
```
is this a typo? `clusterctl v1alpha3 API group`

furkatgofurov7 commented on internal/rancher/provisioning/v1/cluster.go:
```diff
@@ -0,0 +1,34 @@
+package v1
```
should we start using the namings aligned with upstream: `cluster_types.go` in this case?

furkatgofurov7 replied:
Probably as a follow-up, this is not a blocker at all and a real nitty

furkatgofurov7 commented on internal/rancher/provisioning/v1/cluster.go:
```diff
@@ -0,0 +1,34 @@
+}
+
+func init() {
+	SchemeBuilder.Register(&Cluster{}, &ClusterList{})
```
suggestion: there has been a recent effort to avoid reliance on controller runtime schemeBuilder in CAPI (xref: https://github.com/kubernetes-sigs/cluster-api/pull/9045/files) which we could also follow

alexander-demicev replied:
I don't think it's required for our case but it's a good habit to start doing it like it's described in CAPI

alexander-demicev replied:
I don't think it's required for our case but it's a good habit to start doing it like it's described in CAPI

alexander-demicev replied:
I don't think it's required for our case but it's a good habit to start doing it like it's described in CAPI

furkatgofurov7 replied:
I think it is better if we align our code when we are at it and early, rather then leaving the work for future

Danil-Grigorev replied:
Who may "suggest" it and why? I would prefer having an issue describing why we'd likely want to avoid using controller runtime without an API, so this, if makes sense could be tracked separately. This PR has a concrete change in mind.

alexander-demicev replied:
it's done to be able to import API without relying on controller-runtime, in this case, we have these wrappers in an internal package so no other projects can import it. I totally agree with @furkatgofurov7 but I don't want to block this PR because of that.

furkatgofurov7 replied:
Fine as is for now by me

furkatgofurov7 commented on test/e2e/import_test.go:
```diff
@@ -20,29 +20,27 @@ limitations under the License.
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	. "sigs.k8s.io/controller-runtime/pkg/envtest/komega"
 
+	turtelesnaming "github.com/rancher-sandbox/rancher-turtles/util/naming"
```
same nit, can we please check for it repo-wide and fix at once: 
```suggestion
	turtlesnaming "github.com/rancher-sandbox/rancher-turtles/util/naming"
```

furkatgofurov7 reviewed (commented):
Great work @Danil-Grigorev, left some questions/suggestions/nits:

furkatgofurov7 commented:
@Danil-Grigorev FYI, created https://github.com/rancher-sandbox/rancher-turtles/issues/113 to document using structured instead of unstructured

richardcase commented on internal/rancher/provisioning/v1/cluster.go:
```diff
@@ -0,0 +1,34 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Status ClusterStatus `json:"status,omitempty"`
```
nit: we can add comments to the fields as well. But its not needed for this and we can follow up on it.

richardcase reviewed (approved):
Great work @Danil-Grigorev , its good to remove the use of unstructured

richardcase reviewed (approved):
Thanks for the rebase, this lgtm

richardcase commented:
Pinging @alexander-demicev & @furkatgofurov7 to re-approve after rebase.

furkatgofurov7 commented on internal/controllers/import_controller.go:
```diff
@@ -50,10 +48,11 @@ import (
 
-	"github.com/rancher-sandbox/rancher-turtles/internal/rancher"
+	managementv3 "github.com/rancher-sandbox/rancher-turtles/internal/rancher/management/v3"
+	provisioningv1 "github.com/rancher-sandbox/rancher-turtles/internal/rancher/provisioning/v1"
 	"github.com/rancher-sandbox/rancher-turtles/util"
 	turtlesannotations "github.com/rancher-sandbox/rancher-turtles/util/annotations"
-	turtelesnaming "github.com/rancher-sandbox/rancher-turtles/util/naming"
+	turtlesnaming "github.com/rancher-sandbox/rancher-turtles/util/naming"
```
Can we add these 3 to the https://github.com/rancher-sandbox/rancher-turtles/blob/main/.golangci.yml#L41-L62 so that we use the consistent imports in the future?

furkatgofurov7 reviewed (commented):
Last nit:
